### ISSUE TITLE
Moving extension methods on `IResult` (to enable covariance)

### ISSUE THREAD

**Issue by redx177** [vkhorikov/CSharpFunctionalExtensions#530]
Covariance is only working with interfaces. Meaning something like this is possible:

```
public interface IVehicle { }
public class Car : IVehicle { }
public class Bus : IVehicle { }

IResult<IVehicle> carResult1 = Result.Success(new Car());
```

This on the other hand is not possible:
```
Result<IVehicle> carResult2 = Result.Success(new Car());
```

The solution seams simple, just go with `IResult` rather than `Result`.
But all the extension methods ([those](https://github.com/vkhorikov/CSharpFunctionalExtensions/tree/master/CSharpFunctionalExtensions/Result/Methods/Extensions)) are on `Result`, and not on `IResult`. 

Is there a possibility to put all extension methods onto `IResult` rather than `Result`?
The "creation" methods ([those](https://github.com/vkhorikov/CSharpFunctionalExtensions/tree/master/CSharpFunctionalExtensions/Result/Methods)) can stay on `Result`, but they return an `IResult`.
Or am I missing something?

**Comment by hankovich** [comment]
Yes, it's one of the C# limitations. The solution here will be to specify type explicitely like

`var carResult2 = Result.Success<IVehicle>(new Car());`

The problem I see with the extensions to `IResult` are

1. Async extensions. `Task`s/`ValueTask`s are not delegates and interfaces, so `Task<Result<int>>` with not be suitable for extensions accepting `Task<IResult<T>>`, only to extensions accepting `Task<Result<T>>`. The possible solution is to parameterize extensions to accept `Task<TResult<TValue>> where TResult : IResult<TValue>`, but C# will not infer type parameters, so everyone will have to provide them explicitely
2. `IResult` is an interface, so there will be a lot of unnecessary allocations
